python/nnvm/frontend/tensorflow.py

```diff
@@ -0,0 +1,512 @@
+        if op_name == 'broadcast_add' and inputs[0].attr('op_name') in conv_ops:
+            # TODO: remove hard coded infershape
+            inputs[1] = _sym.expand_dims(inputs[1], axis=axis, num_newaxis=2)
+        return get_nnvm_op(op_name)(*inputs)
+    return _impl
+
+def _pooling(name):
+    def _impl(inputs, attr, params):
+
+        attr['data_format'] = attr['data_format'].decode("utf-8")
+
+        if attr['data_format'] == 'NHWC':
+            attr['kernel_shape'] = (attr['ksize'][1], attr['ksize'][2])
+        if attr['data_format'] == 'NCHW':
+            attr['kernel_shape'] = (attr['ksize'][2], attr['ksize'][3])
+
+        # Fix strides
+        attr['strides'] = (attr['strides'][1], attr['strides'][2])
+
+        # Fix padding
+        input_shapes = _get_input_shapes(attr)
+        attr['padding'] = attr['padding'].decode("utf-8")
+
+        if attr['padding'] == 'VALID':
+            attr['padding'] = [0, 0]
+        elif attr['padding'] == 'SAME':
+            stride_h, stride_w = attr['strides']
+            kernel_h, kernel_w = attr['kernel_shape']
+            in_h = input_shapes[0][1]
+            in_w = input_shapes[0][2]
+            pad_t = _get_pad(in_h, kernel_h, stride_h)
+            pad_l = _get_pad(in_w, kernel_w, stride_w)
+            attr['padding'] = [pad_t, pad_l]
+        else:
+            raise TypeError("Unsupported padding type : {}".format(attr['padding']))
+
+        return AttrCvt(
+            op_name=_dimension_picker(name),
+            transforms={
+                'kernel_shape':'pool_size',
+                'data_format':'layout'},
+            ignores=['ksize'],
+            extras={'ceil_mode': False},
+            custom_check=_dimension_constraint())(inputs, attr)
+    return _impl
+
+def _conv():
+    def _impl(inputs, attr, params):
+        attr['data_format'] = attr['data_format'].decode("utf-8")
+
+        # Extract kernel shape from params
+        conv_param_weights = params[inputs[1].list_output_names()[0]]
+
+        if attr['data_format'] == 'NHWC':
+            attr['kernel_shape'] = (conv_param_weights.shape[0], conv_param_weights.shape[1])
+            attr['channels'] = conv_param_weights.shape[3]
+            if 'dilations' in attr:
+                attr['dilations'] = (attr['dilations'][0], attr['dilations'][1])
+        if attr['data_format'] == 'NCHW':
+            attr['kernel_shape'] = (conv_param_weights.shape[2], conv_param_weights.shape[3])
+            attr['channels'] = conv_param_weights.shape[1]
+            if 'dilations' in attr:
+                attr['dilations'] = (attr['dilations'][2], attr['dilations'][3])
+
+        # Fix strides
+        attr['strides'] = (attr['strides'][1], attr['strides'][2])
+
+        # Fix padding
+        input_shapes = _get_input_shapes(attr)
+        attr['padding'] = attr['padding'].decode("utf-8")
+
+        if attr['padding'] == 'VALID':
+            attr['padding'] = [0, 0]
+        elif attr['padding'] == 'SAME':
```

Comment:
I am afraid SAME padding is not correctly converted here.
Can you add a test with input_h=32, kernel_h=3, stride_h=2?

Reply:
@Huyuwei Aha, this keras's implementation is interesting. But how to we handle the workload / schedule part? Because in our workload / schedule part does only concern VALID padding and only care top / left. see: https://github.com/dmlc/tvm/blob/9175b47c16691824b1a54aa26dc1ddb21c981612/topi/python/topi/nn/conv2d.py#L146 Though Keras's implementation solve SAME padding, but not solve pass this padding information to the scheduling part. This is why I discuss with @tqchen in https://discuss.tvmlang.org/t/tvm-seems-conv2d-has-relate-big-floating-point-error-when-to-from-coreml/145. @tqchen suggest extending Conv2D's operator's param(maybe include pooling) form [top, left] to [top, left, bottom, right]. So I want to know this Keras's implementation. how to pass the 4 padding information to scheduling? If we can solve this problem, we don't need to extend operator's param.

Reply:
Initially I borrowed the pad_pair calculation from keras which I replaced with tensorflow reference.
Both work fine with InceptionV1 as long as I don't use pad operator before convolution, but pass pad_t, pad_l to conv2d.

I see the below error with pad operator. I am investigating it.

nnvm._base.NNVMError: Error in operator conv2d2: [14:09:41] src/top/nn/convolution.cc:84: Operator conv2d(use_bias=False, strides=(1, 1), channels=64, layout=NHWC, padding=[0, 0], kernel_size=(3, 3), kernel_layout=HWIO, name=conv2d2) expects weight's shape to be [3,3,34,64], but got [3,3,32,64].

Reply:
Above error due to different order of padding values for NHWC. 

Now the padding scheme for TF front end follows tensorflow reference with pad operator before convolution and it work fine with InceptionV1.

Reply:
@FrozenGene Now keras frontend inserts a pad operator before conv to handle SAME padding, and I expect the pad operator to be fused. Consider this sequence: extra_pad_1, pad_1, conv_1, relu_1, extra_pad_2, pad_2, conv2... relu_1 and extra_pad_2 will all be fused into conv_1. But I just noticed that extra_pad_2 won't be fused in the generated code. So this pad operator brings overhead.

I agree that we can extend conv2d's padding param to support [top, left, bottom, right]. It will make the conversion much easier and clearer. The pad API in keras may be helpful: https://keras.io/layers/convolutional/#zeropadding2d

Reply:
@Huyuwei Yes, but it will need much work in the scheduling part, because we don't consider bottom / right padding value in our previous design, I am working on it. And we will also need to extend conv2d_transpose / max_pool2d / avg_pool2d / global_max_pool2d / global_avg_pool2d. i.e. we need to extend all operators of Level 2 Convolutions in this table: http://nnvm.tvmlang.org/top.html#overview-of-operators

Reply:
Another issue I came across with padding on pool operation while doing test cases for tensorflow frontend https://github.com/dmlc/nnvm/issues/486

I think we could move this topic to discussion forum.

Reply:
@srkreddy1238 @Huyuwei I have extended padding completely and can have the same result as Tensorflow / CoreML now. However, I keep the implementation of padding [top, left], which is useful for the compatibility of old code. For example, we have unittest test_infer_shape using old behaviour. But if user specify [top, left, bottom, right], we go to the new behaviour. Another reason I keep the implementation is I find many framework(keras, MXNet and so on) are using padding[top, left], I don't have time to test every framework(Especially MXNet, I find its official documentation says that his padding parameter is only[top, left] too). But I think we should only have padding[top, left, bottom, right]. Do you have any ideas or suggestions?

Reply:
Sounds good. Can you PR it to review ?

Reply:
👍 . I will help to check tensorflow once you PR it.